### ISSUE TITLE
Migrate command - [merged]

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#1921]
In GitLab by @dbeltrankyl on May 16, 2024, 14:37

_Merges 4.1.7-migrate -> master_

Migrate stuff updated from 4.X.X branch

Working on

#947

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on May 16, 2024, 14:39

mentioned in merge request ces/autosubmit4-config-parser!5

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on May 16, 2024, 14:40

https://earth.bsc.es/gitlab/ces/autosubmit4-config-parser/-/merge_requests/5 <- relevant update in order to migrate to work properly

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on May 22, 2024, 14:57

added 1 commit

<ul><li>68f0d147 - Tests pending</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/433/diffs?diff_id=25531&start_sha=c06f4d41b5abba6c37d4d9d142e223a2b096dafd)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on May 23, 2024, 08:36

added 2 commits

<ul><li>c202fda8 - progress bar</li><li>0069d215 - migrate jobdata</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/433/diffs?diff_id=25545&start_sha=68f0d1471b3f6baa87c52dce8932201f5ec4f209)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 5, 2024, 17:16

added 4 commits

<ul><li>7f8b895d - (tests wip)</li><li>8de62ffe - tests</li><li>910a6f7f - added temporal db for pytests at class scope</li><li>2df61361 - fixed few things, functional test working</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/433/diffs?diff_id=26358&start_sha=0069d2150eb87b9f278762e333a5949faf0fc9b6)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 6, 2024, 10:07

added 1 commit

<ul><li>edffe0b5 - fixed few things, functional test working</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/433/diffs?diff_id=26370&start_sha=2df613610f2ad9cc86f26d9fa1e40c49e5b09dc6)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 6, 2024, 15:48

added 1 commit

<ul><li>1a65091d - improved tests, added if file exists</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/433/diffs?diff_id=26400&start_sha=edffe0b505c0454e745b7a5602d6c83971337e6f)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 6, 2024, 16:11

added 1 commit

<ul><li>3cebf151 - working</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/433/diffs?diff_id=26404&start_sha=1a65091d6a32499af7d5b4a4522d446f21769f6f)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 7, 2024, 11:01

added 1 commit

<ul><li>f2125e43 - improved tests, added if file exists</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/433/diffs?diff_id=26433&start_sha=3cebf15172a6fbeb2e8979e00a3e15545dd8fa6d)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 7, 2024, 12:46

added 1 commit

<ul><li>7d6de1d2 - Improved tests and check if conf is ok</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/433/diffs?diff_id=26440&start_sha=f2125e433003872f8045a1e5d68ccf8bac815a87)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 7, 2024, 13:03

added 1 commit

<ul><li>d66689e4 - Added docs</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/433/diffs?diff_id=26442&start_sha=7d6de1d253248665181b308f9f62695bd3ea2f33)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 7, 2024, 13:54

Hello @kinow 

I'm having a weird issue with the pipeline and when running all the tests at once. 

Pycharm->right click in unit_folder -> run pytests in unit # Migrate tests doesn't work

Pycharm->right click in test_migrate.py -> run pytests test_migrate.py # Migrate tests working

Debugging it I found that: 


```python
@ staticmethod
def install():
    """
    Creates a new database instance for autosubmit at the configured path

    """
    if not os.path.exists(BasicConfig.DB_PATH): # BasicConfig.DB_PATH points to the tmp folder/tests.db and it doesn't exists
        Log.info("Creating autosubmit database...")
        qry = resource_string('autosubmit.database', 'data/autosubmit.sql').decode(locale.getlocale()[1])
        #qry = importlib.resources.read_text('autosubmit.database', 'data/autosubmit.sql').decode(locale.getlocale()[1])
        if not create_db(qry):
            raise AutosubmitCritical("Can not write database file", 7004)
        Log.result("Autosubmit database created successfully")
    else:
        raise AutosubmitCritical("Database already exists.", 7004)
    return True
```

It is crashing when running all folders because it tells me that the (BasicConfig.DB_PATH) already exists. I've checked with the debugger that all paths are correct, BasicConfig.DB_PATH /tmp/pytest-of-dbeltran/pytest-#number/migrate_tests0/tests.db

Manually, I went to that folder and, indeed, the tests.db still doesn't exist, 

I've tried to change the scope to session and class, but the same thing happens. 

Do you know why os.path.exists is lying and how to fix it?

Edit: 
![Selection_145](https://earth.bsc.es/gitlab/es/autosubmit/uploads/48bb1e28583fd645542749bd0a7f24c1/Selection_145.png)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 7, 2024, 14:26

Found the issue, 

The os.path.exists is mocked forever from other tests. Do you know if I can un mock something?

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 7, 2024, 14:39

Had to comment these three tests that are doing nasty stuff

```
    # def test_exists_completed_file_then_sets_status_to_completed(self):
    #     # arrange
    #     exists_mock = Mock(return_value=True)
    #     sys.modules['os'].path.exists = exists_mock
    #
    #     # act
    #     self.job.check_completion()
    #
    #     # assert
    #     exists_mock.assert_called_once_with(os.path.join(self.job._tmp_path, self.job.name + '_COMPLETED'))
    #     self.assertEqual(Status.COMPLETED, self.job.status)

    # def test_completed_file_not_exists_then_sets_status_to_failed(self):
    #     # arrange
    #     exists_mock = Mock(return_value=False)
    #     sys.modules['os'].path.exists = exists_mock
    #
    #     # act
    #     self.job.check_completion()
    #
    #     # assert
    #     exists_mock.assert_called_once_with(os.path.join(self.job._tmp_path, self.job.name + '_COMPLETED'))
    #     self.assertEqual(Status.FAILED, self.job.status)
    # def test_when_database_already_exists_then_is_not_initialized_again(self):
    #     sys.modules['os'].path.exists = MagicMock(return_value=True)
    #     connection_mock = MagicMock()
    #     cursor_mock = MagicMock()
    #     cursor_mock.side_effect = Exception('This method should not be called')
    #     connection_mock.cursor = MagicMock(return_value=cursor_mock)
    #     original_connect = sys.modules['sqlite3'].connect
    #     sys.modules['sqlite3'].connect = MagicMock(return_value=connection_mock)
    #     DbManager('dummy-path', 'dummy-name', 999)
    #     connection_mock.cursor.assert_not_called()
    #     sys.modules['sqlite3'].connect = original_connect

```

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 7, 2024, 14:39

added 1 commit

<ul><li>f2a046e2 - Disabled tests that are messing with the loaded sys.module</li></ul>

[Compare with previous version](/gitlab/es/autosubmit/-/merge_requests/433/diffs?diff_id=26462&start_sha=d66689e4d02308dc56c025acccae08d41e09cf51)

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 7, 2024, 14:41

marked this merge request as **ready**

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 7, 2024, 14:59

Now the pipeline is not working due it is using "nose" but since it will be changed in another branch I'll leave it as it is now

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 7, 2024, 15:35

mentioned in commit 4aa9a2a493fd3de7032d8911c3eb9407c75c6110

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 7, 2024, 15:37

mentioned in commit d4bebddbacf8787efba5894ecba2fd723ff9988d

**Comment by kinow** [comment]
In GitLab by @mcastril on Jun 17, 2024, 20:19

Dani, in the weekly meeting I remember saying that we should not merge this now as it would be a problem for others working in parallel branches, and also because it's something quite extensive and not prioritary. Most of the code will need a similar refactor and it'll be probably better to do it at the same time, with more calm.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 18, 2024, 08:23

Hello @mcastril ,

This was already merged and working in v4.1.8. What we discussed was the start conditions one https://earth.bsc.es/gitlab/es/autosubmit/-/issues/1316

**Comment by kinow** [comment]
I think this was discussed a few meetings ago, before the release. I mentioned we had an issue for that with another possible solution #973

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 18, 2024, 08:42

Yes, but that was when we had to do the splits stuff, etc, the last meeting, the issue was with the   start conditions to make the configuration simpler ( I have yet to revise the feedback )

**Comment by kinow** [comment]
In GitLab by @mcastril on Jun 18, 2024, 09:29

Sorry if my message was misleading. I didn't mean the last weekly meeting but a previous one, where you informed us that you had taken advantage of the migrate developments to refactor this part of the code.

Bruno and Luiggi said that they were working in parallel in the PostgreSQL development, and that it would be difficult for them to catch up if there. And then I said that, as more refactors will be needed anyway, we should postpone this to later in Summer or even for AS5.

Maybe this rings you a bell now

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 18, 2024, 11:33

> this brings a big change to that file central to all the developments.

I don't think that it is a big change in the autosubmit.py. I reverted the big changes ( that are the moves of some functions non-migrate functions) based on our meeting because It is what I thought we accorded. I did not ask for review due Bruno being busy in Edito 

> But I don't see changes in `job_utils.py`... https://earth.bsc.es/gitlab/es/autosubmit/-/commits/master/autosubmit/job/job_utils.py What do you mean exactly?

I meant the utils.helpers, sorry most of the stuff touched there are additions, which are functions that are in the autosubmit.py, and the deletions are non-used imports and normalized the code with the pycharm tool ( that maybe I shouldn't have used the tool at that time).

What I understood is that the big change was what I was doing in which I was moving all the autosubmit.py functions that migrate used in different files, but after the meeting, I reverted it. I don't see a problem with moving the migrate-related ones because they won't be touched outside that branch, but I may be wrong. 


About autosubmit.helper, I performed the pycharm tool there by mistake didn't notice.

I think it was misunderstood but I will take the advice in the future.

**Comment by kinow** [comment]
In GitLab by @mcastril on Jun 18, 2024, 14:45

> I don't think that it is a big change in the autosubmit.py. I reverted the big changes ( that are the moves of some functions non-migrate functions) based on our meeting because It is what I thought we accorded. I did not ask for review due Bruno being busy in Edito

> What I understood is that the big change was what I was doing in which I was moving all the autosubmit.py functions that migrate used in different files, but after the meeting, I reverted it. I don't see a problem with moving the migrate-related ones because they won't be touched outside that branch, but I may be wrong.

OK @dbeltrankyl, I think I didn't realize of the changes that were removed from the MR

I remember the distribution in different files, and it's true that it's not part of this MR

> I don't see a problem with moving the migrate-related ones because they won't be touched outside that branch, but I may be wrong.

I was concerned about whether it was a problem for Bruno or Luiggi. I cannot judge the impact better than you do from my position. Anyway, thanks for taking the advice and reducing the potential changes. We'll discuss how to improve the development workflow now that the team is increasing again. It's something that will take some effort to us but I am sure it'll pay off.
